pallets/messages/src/lib.rs

```diff
@@ -344,28 +351,39 @@ impl<T: Config> Pallet<T> {
 		schema_id: SchemaId,
 		current_block: BlockNumberFor<T>,
 	) -> Result<bool, DispatchError> {
-		<Messages<T>>::try_mutate(
-			current_block,
-			schema_id,
-			|existing_messages| -> Result<bool, DispatchError> {
-				// first message for any schema_id is going to trigger an event
-				let need_event = existing_messages.len() == 0;
-				let index = BlockMessageIndex::<T>::get();
-				let msg = Message {
-					payload, // size is checked on top of extrinsic
-					provider_msa_id,
-					msa_id,
-					index,
-				};
+		let mut metadata = MessageBlockMetadata::<T>::get();
+		ensure!(
+			u32::from(metadata.total_index) < T::MaxMessagesPerBlock::get(),
+			Error::<T>::TooManyMessagesInBlock
+		);
+		let mut found = false;
+		let mut current_index: MessageIndex = 0;
+		for schema_count in metadata.schema_counts.iter_mut() {
+			if schema_count.schema_id == schema_id {
+				found = true;
+				current_index = schema_count.count;
+				schema_count.count = schema_count.count.saturating_add(1);
+			}
+		}
+		if !found {
+			metadata
+				.schema_counts
+				.try_push(SchemaCount { count: 1, schema_id })
+				.map_err(|_| Error::<T>::TooManyMessagesInBlock)?;
```

Comment:
The whitelist would still remain what we would still have an index and that index would be a counter for each message. I think this would work because we never access messages on the chain. In the case that we did would provide all the keys required to access that message. 

There are tradeoffs in that we might be pushing logic to off-chain, however, since the messages themselves contain an index relative ordering for each schema message is still possible.

Reply:
Applied